cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/lineage/field/FieldLineageDataset.java

```diff
@@ -0,0 +1,254 @@
+   * @param start start time in seconds
+   * @param end end time in seconds
+   * @return {@link Set} of {@link FieldLineageInfo} instances
+   */
+  public Set<FieldLineageInfo> getFieldLineageInfosInRange(EndPoint endPoint, long start, long end) {
+    Set<Long> checksums = getChecksumsInRange(getScanStartKey(endPoint, end, BACKWARD_LINEAGE_MARKER),
+                                              getScanEndKey(endPoint, start, BACKWARD_LINEAGE_MARKER));
+
+    Set<FieldLineageInfo> infos = new HashSet<>();
+    for (long checksum : checksums) {
+      infos.add(getFieldLineageInfo(checksum));
+    }
+    return infos;
+  }
+
+  private FieldLineageInfo getFieldLineageInfo(long checksum) {
+    byte[] rowKey = getChecksumRowKey(checksum);
+    byte[] bytes = table.get(rowKey, RAW_FORMAT_COLS_BYTE);
+    return GSON.fromJson(Bytes.toString(bytes), FieldLineageInfo.class);
+  }
+
+  private byte[] getScanStartKey(EndPoint endPoint, long end, char lineageDirectionMarker) {
+    // time is inverted, hence we need to have end time in start key.
+    return getScanKey(endPoint, end, lineageDirectionMarker);
+  }
+
+  private byte[] getScanEndKey(EndPoint endPoint, long start, char lineageDirectionMarker) {
+    // time is inverted, hence we need to have start time in end key.
+    return getScanKey(endPoint, start, lineageDirectionMarker);
+  }
+
+  private Set<Long> getChecksumsInRange(byte[] startKey, byte[] endKey) {
+    Set<Long> checksums = new HashSet<>();
+    try (Scanner scanner = table.scan(startKey, endKey)) {
+      Row row;
+      while ((row = scanner.next()) != null) {
+        if (LOG.isTraceEnabled()) {
+          LOG.trace("Got row key = {}", Bytes.toString(row.getRow()));
+        }
+
+        checksums.add(Bytes.toLong(row.get(CHECKSUM_COLS_BYTE)));
+      }
+    }
+    return checksums;
+  }
+
+  private byte[] getScanKey(EndPoint endPoint, long time, char lineageDirectionMarker) {
+    MDSKey.Builder builder = new MDSKey.Builder();
+    builder.add(lineageDirectionMarker);
+    addEndPoint(builder, endPoint);
+    builder.add(invertTime(time));
+    return builder.build().getKey();
+  }
+
+  private void addOperationReferenceRecord(ProgramRunId programRunId, EndPoint endPoint, long checksum,
+                                           char lineageDirectionMarker) {
+    byte[] bytes = getProgramRunKey(lineageDirectionMarker, programRunId, endPoint);
+    table.put(bytes, CHECKSUM_COLS_BYTE, Bytes.toBytes(checksum));
+  }
+
+  private byte[] getChecksumRowKey(long checksum) {
+    MDSKey.Builder builder = new MDSKey.Builder();
+    builder.add("c");
+    builder.add(checksum);
+    return builder.build().getKey();
+  }
+
+  private byte[] getProgramRunKey(char lineageDirectionMarker, ProgramRunId programRunId, EndPoint endPoint) {
+    long invertedStartTime = getInvertedStartTime(programRunId);
+    MDSKey.Builder builder = new MDSKey.Builder();
+    builder.add(lineageDirectionMarker);
```

Comment:
MDSKey.Builder does not have a method for add(char), so this probably translates to an integer, using 4 bytes instead of 1? Perhaps think about adding a method for a single byte (because that is what you actually need, even char is 16 bits that is 2 bytes).

Reply:
I also think you should be consistent in the way you construct keys. Above you add "c" as a string, here you add a char.